src/sentry/debug_files/artifact_bundles.py

```diff
@@ -34,14 +50,374 @@
+) -> FlatFileIndexingState:
+    with ArtifactBundleArchive(artifact_bundle.file.getfile(), build_memory_map=True) as archive:
+        identifiers = [identifier]
+
+        # In case we have an identifier that is indexed by release and the bundle has debug ids, we want to perform
+        # indexing on also the debug ids bundle.
+        if identifier.is_indexing_by_release() and archive.has_debug_ids():
+            identifiers.append(identifier.to_indexing_by_debug_id())
+
+        # Before starting, we have to make sure that no other identifiers are being indexed.
+        if not set_flat_files_being_indexed_if_null(identifiers):
+            return FlatFileIndexingState.CONFLICT
+
+        # For each identifier we now have to compute the index.
+        for identifier in identifiers:
+            index = FlatFileIndex.build(FlatFileIndexStore(identifier))
+            if index is None:
+                return FlatFileIndexingState.ERROR
+
+            # Once the in-memory index is built, we can merge it with the incoming bundle.
+            if not index.merge(artifact_bundle, archive):
+                return FlatFileIndexingState.ERROR
+
+            # At the end we want to save the newly created index.
+            if not index.save():
+                return FlatFileIndexingState.ERROR
+
+    return FlatFileIndexingState.SUCCESS
+
+
+class BundleMeta(TypedDict):
+    id: int
+    timestamp: datetime
+
+
+Bundles = List[BundleMeta]
+FilesByUrl = Dict[str, List[int]]
+FilesByDebugID = Dict[str, List[int]]
+
+
+class FlatFileIndexStore:
+    def __init__(self, identifier: FlatFileIdentifier):
+        self.identifier = identifier
+
+    def load(self) -> Optional[Dict[str, Union[Bundles, FilesByUrl, FilesByDebugID]]]:
+        try:
+            flat_file_index = ArtifactBundleFlatFileIndex.objects.get(
+                project_id=self.identifier.project_id,
+                release_name=self.identifier.release,
+                dist_name=self.identifier.dist,
+            ).load_flat_file_index()
+
+            return self._deserialize_json_index(flat_file_index)
+        except ArtifactBundleFlatFileIndex.DoesNotExist:
+            return None
+
+    @classmethod
+    def _deserialize_json_index(
+        cls, json_index: Dict[str, Any]
+    ) -> Dict[str, Union[Bundles, FilesByUrl, FilesByDebugID]]:
+        deserialized_index: Dict[str, Union[Bundles, FilesByUrl, FilesByDebugID]] = {}
+
+        if "bundles" in json_index:
+            deserialized_index["bundles"] = cls._deserialize_bundles(json_index["bundles"])
+
+        if "files_by_url" in json_index:
+            deserialized_index["files_by_url"] = json_index["files_by_url"]
+
+        if "files_by_debug_id" in json_index:
+            deserialized_index["files_by_debug_id"] = json_index["files_by_debug_id"]
+
+        return deserialized_index
+
+    @classmethod
+    def _deserialize_bundles(cls, bundles: List[Dict[str, Union[int, str]]]) -> Bundles:
```

Comment:
These to/from JSON methods could as well just be implemented on `FlatFileIndex`.